src/lti_provider/validator.py

```diff
@@ -0,0 +1,194 @@
+
+        This method must allow the use of a dummy client_key value.
+        Fetching the secret using the dummy key must take the same amount of
+        time as fetching a secret for a valid client::
+
+        Args:
+            client_key: The client/consumer key.
+            request: The calling request.
+        """
+
+        try:
+            passport = LTIPassport.objects.get(
+                oauth_consumer_key=client_key, is_enabled=True
+            )
+            return passport.shared_secret
+        except LTIPassport.DoesNotExist:
+            return "dummy_client_sec_123456"
+
+    def validate_client_key(self, client_key, request):
+        """Validates that supplied client key is a registered and valid client.
+
+        Args:
+            client_key: The client/consumer key.
+            request: The calling request
+
+        Returns:
+            bool: True if the client key is registered and valid
+        """
+        return LTIPassport.objects.filter(
+            oauth_consumer_key=client_key, is_enabled=True
+        ).exists()
+
+    # pylint: disable=too-many-arguments
+    def validate_timestamp_and_nonce(
+        self,
+        client_key: str,
+        timestamp: str,
+        nonce: str,
+        request,
+        request_token=None,
+        access_token=None,
+    ):
+        """Validates that the nonce has not been used before
+
+        Per `Section 3.3`_ of the OAuth 1.0 spec.
+
+        "A nonce is a random string, uniquely generated by the client to allow
+        the server to verify that a request has never been made before and
+        helps prevent replay attacks when requests are made over a non-secure
+        channel.  The nonce value MUST be unique across all requests with the
+        same timestamp, client credentials, and token combinations."
+
+        .. _`Section 3.3`: https://tools.ietf.org/html/rfc5849#section-3.3
+
+        One of the first validation checks that will be made is for the validity
+        of the nonce and timestamp, which are associated with a client key and
+        possibly a token. If invalid then immediately fail the request
+        by returning False. If the nonce/timestamp pair has been used before and
+        you may just have detected a replay attack. Therefore it is an essential
+        part of OAuth security that you not allow nonce/timestamp reuse.
+        Note that this validation check is done before checking the validity of
+        the client and token.::
+
+        Args:
+            client_key: The client/consumer key.
+            timestamp: The ``oauth_timestamp`` request parameter.
+            nonce: The ``oauth_nonce`` request parameter.
+            request: The calling request
+            request_token: Request token string, if any.
+            access_token: Access token string, if any.
+
+        Returns:
+            bool: True if the timestamp and nonce has not been used before
+        """
+
+        cache_timeout = 3600
```

Comment:
cache timeout is one of the parameters when declaring a cache backend in Django so maybe you could use it?

Reply:
Since this is a security related option, I don't feel comfortable with reusing a cache backend option that has been designed for performance. A too small value can cause issues, and a too big value can lead us to accept LTI request signed long time ago.

 I would prefer to make an option in `lti_provider` settings, instead of this hardcoded value, WDYT ?

Reply:
ok let's leave it like this then.

Reply:
Hmm then I think we need a test to confirm that the cache timeout defined in the backend is overriden by what you pass. Otherwise your security issues may be comprised?

Reply:
As described in [Django's documentation](https://docs.djangoproject.com/en/2.2/topics/cache/#cache-arguments), the `TIMEOUT` option in the backend configuration is a default value, and each [api call](https://docs.djangoproject.com/en/2.2/topics/cache/#basic-usage) allows to set a custom timeout.

Why testing this if it's well documented and this is the responsibility of each cache backend ?